Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
commit c1379c25e04a051e789c55be3ce730bbbd51c905
Author: Thomas Gleixner <tglx@linutronix.de>
Date:   Sun May 18 22:17:59 2008 +0200

    x86: disable TSC for sched_clock() when calibration failed
    
    When the TSC calibration fails then TSC is still used in
    sched_clock(). Disable it completely in that case.
    
    Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
    Cc: stable@kernel.org

---
 arch/x86/kernel/tsc_32.c |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

Index: linux-2.6.25.4-rt4/arch/x86/kernel/tsc_32.c
===================================================================
--- linux-2.6.25.4-rt4.orig/arch/x86/kernel/tsc_32.c	2008-05-29 09:46:00.000000000 -0400
+++ linux-2.6.25.4-rt4/arch/x86/kernel/tsc_32.c	2008-05-29 09:46:01.000000000 -0400
@@ -399,8 +399,14 @@ void __init tsc_init(void)
 	cpu_khz = calculate_cpu_khz();
 	tsc_khz = cpu_khz;
 
-	if (!cpu_khz)
+	if (!cpu_khz) {
+		/*
+		 * We need to disable the TSC completely in this case
+		 * to prevent sched_clock() from using it.
+		 */
+		tsc_disabled = 1;
 		goto out_no_tsc;
+	}
 
 	printk("Detected %lu.%03lu MHz processor.\n",
 				(unsigned long)cpu_khz / 1000,
